File: src/main/java/com/github/tanokun/tanorpg/game/entity/ActiveEntity.java
package com.github.tanokun.tanorpg.game.entity;

import com.github.tanokun.tanorpg.game.entity.base.ObjectEntity;
import com.github.tanokun.tanorpg.player.status.StatusType;
import org.bukkit.entity.Entity;

public class ActiveEntity {
    private final ObjectEntity objectEntity;

    private final Entity entity;

    private int hasHP;

    public ActiveEntity(ObjectEntity objectEntity, Entity entity) {
        this.objectEntity = objectEntity;
        this.entity = entity;
        this.hasHP = objectEntity.getStatusMap().getStatus(StatusType.HP);
    }

    public ObjectEntity getObjectEntity() {
        return objectEntity;
    }

    public int getHasHP() {
        return hasHP;
    }

    public int setHasHP(int hasHP) {
        this.hasHP = hasHP;
        return this.hasHP;
    }
}

File: src/main/java/com/github/tanokun/tanorpg/game/entity/ActiveEntity.kt
package com.github.tanokun.tanorpg.game.entity

import com.github.tanokun.tanorpg.TanoRPG
import com.github.tanokun.tanorpg.damage.damageByNull
import com.github.tanokun.tanorpg.player.status.StatusType
import org.bukkit.attribute.Attributable
import org.bukkit.attribute.Attribute
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.metadata.FixedMetadataValue
import java.util.*
import kotlin.math.roundToInt

open class ActiveEntity(val objectEntity: ObjectEntity, val entity: Entity) {
    var health: Int = (entity as Attributable).getAttribute(Attribute.GENERIC_MAX_HEALTH)!!.value.roundToInt()

    fun removeHealth(value: Int, player: Player): Int {
        val r = health

        health -= value
        if (health <= 0) {
            entity.setMetadata("isDead", FixedMetadataValue(TanoRPG.plugin, player))
            damageByNull(entity, 10000000000.0)
        }

        return r
    }

    val exp = HashMap<UUID, Double>()

    val oneExp: Double
        get() {
            val exp = objectEntity.exp.toDouble()
            val hp = (entity as Attributable).getAttribute(Attribute.GENERIC_MAX_HEALTH)?.value ?: return 0.0
            return exp / hp
        }
}